rescheduler/rescheduler.go

```diff
@@ -77,8 +88,247 @@ func main() {
+
+						nodes, err := nodeLister.List()
+						if err != nil {
+							glog.Errorf("Failed to list nodes: %v", err)
+							continue
+						}
+
+						node := findNodeForPod(kubeClient, predicateChecker, nodes, pod)
+						if node == nil {
+							glog.Errorf("Pod %s can't be scheduled on any existing node.", podId(pod))
+							continue
+						}
+						glog.Infof("Trying to place the pod on node %v", node.Name)
+
+						err = prepareNodeForPod(kubeClient, predicateChecker, node, pod)
+						if err != nil {
+							glog.Warningf("%+v", err)
+						} else {
+							podsBeingProcessed.Add(pod)
+							go waitForScheduled(podsBeingProcessed, pod)
+						}
+					}
+				}
+
+				releaseAllTaints(kubeClient, nodeLister, podsBeingProcessed)
 			}
 		}
 	}
 }
+
+func waitForScheduled(podsBeingProcessed *podSet, pod *kube_api.Pod) {
+	// TODO(piosz): improve this
+	glog.Infof("Waiting for pod %s to be scheduled", podId(pod))
+	time.Sleep(10 * time.Minute)
+	podsBeingProcessed.Remove(pod)
+}
+
+func createKubeClient(flags *flag.FlagSet, inCluster bool) (*kube_client.Client, error) {
+	if inCluster {
+		return kube_client.NewInCluster()
+	}
+
+	clientConfig := kubectl_util.DefaultClientConfig(flags)
+	config, err := clientConfig.ClientConfig()
+	if err != nil {
+		fmt.Errorf("error connecting to the client: %v", err)
+	}
+	return kube_client.NewOrDie(config), nil
+}
+
+func releaseAllTaints(client *kube_client.Client, nodeLister *kube_utils.ReadyNodeLister, podsBeingProcessed *podSet) {
+	nodes, err := nodeLister.List()
+	if err != nil {
+		glog.Warningf("Cannot release taints - error while listing nodes: %v", err)
+		return
+	}
+
+	for _, node := range nodes {
+		taints, err := kube_api.GetTaintsFromNodeAnnotations(node.Annotations)
+		if err != nil {
+			glog.Warningf("Error while getting Taints for node %v: %v", node.Name, err)
+			continue
+		}
+
+		newTaints := make([]kube_api.Taint, 0)
+		for _, taint := range taints {
+			if taint.Key == criticalAddonsOnly && podsBeingProcessed.HasId(taint.Value) {
+				glog.Infof("Releasing taint %+v on node %v", taint, node.Name)
+			} else {
+				newTaints = append(newTaints, taint)
+			}
+		}
+
+		if len(newTaints) != len(taints) {
+			taintsJson, err := json.Marshal(newTaints)
+			if err != nil {
+				glog.Warningf("Error while releasing taints on node %v: %v", node.Name, err)
+				continue
+			}
+
+			node.Annotations[kube_api.TaintsAnnotationKey] = string(taintsJson)
+			_, err = client.Nodes().Update(node)
+			if err != nil {
+				glog.Warningf("Error while releasing taints on node %v: %v", node.Name, err)
+			} else {
+				glog.Infof("Successfully released all taints on node %v", node.Name)
+			}
+		}
+	}
+}
+
+func prepareNodeForPod(client *kube_client.Client, predicateChecker *ca_simulator.PredicateChecker, node *kube_api.Node, criticalPod *kube_api.Pod) error {
+	// This taint will be removed if this function will return an error.
```

Comment:
I think you mean "The caller of this function must remove the taint if this function returns error" ?
(There's no code that I can see in this function that removes it if this function returns error.)